visualizer/main.py

```diff
@@ -0,0 +1,399 @@
+
+        return arbiters
+
+    #find and register all consumers and assign a color to each unique one
+
+    def assign_colors_to_consumers(self, sim, arbiters):
+
+        ctcr = {}
+        count = 0
+
+        l_colors = list(self.l_c_dict.keys())
+        d_colors = list(self.d_c_dict.keys())
+
+        for dp in sim:
+            if isinstance(dp, ConsumerDataPoint) and dp.consumer_id not in ctcr:
+
+                l_color = l_colors[count]
+                d_color = d_colors[count]
+                ctcr[dp.consumer_id] = (l_color, d_color, count, dp.cycle_number)
+                count += 1
+                pass_con_Id = dp.consumer_id
+
+                self.create_var_for_con(ctcr, pass_con_Id)
+
+        return ctcr
+
+    #print the arb variables
+    #in order to do so, require to have an updated self.offset_row
+
+    def create_var_for_arb(self, arbiters, pass_arb_Id):
+        row = self.offset_row
+
+        self.requests.append(input_data_into_cell(row, 0, pass_arb_Id))
+
+        self.offset_row += 1
+
+    #print the arb variables
+    #in order to do so, require to have an updated self.offset_row
+
+    def create_var_for_con(self, ctcr, pass_con_Id):
+        row = self.offset_row
+
+
+        self.requests.append(input_data_into_cell(row, 0, pass_con_Id))
+
+        self.offset_row += 1
+
+    #get ready the info and the updated self.offset_row
+
+    def info_renderer(self, count):
+        self.print_config_data(count)
+        #self.define_clock_length()
+        self.requests.append(input_data_into_cell_merged(self.offset_row, self.offset_row + 1, 1, 25))
+        self.offset_row += 1
+        self.requests.append(input_data_into_cell(self.offset_row, 1, "0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18"))
+        self.offset_row += 1
+        self.start_of_arb_in_sim = self.offset_row
+
+    #print the config data
+
+    def define_clock_length(self, ctcr):
+        for sim in self.sims:
+            for dp in sim:
+                #find the highest nunmber in the cycle_number
+                pass
+
+    def print_config_data(self, count):
+        for sim in self.sims:
+            config = self.config[count]
+            self.requests.append(input_data_into_cell(self.offset_row, self.offset_row + 1, config))
+
+
+    #how the process for each simulation looks like
+
+    def render_simulation(self, sim, count):
+
+        self.info_renderer(count)
+
+        arbiters = self.assign_arbiters(sim)
+        ctcr = self.assign_colors_to_consumers(sim, arbiters)
+
+        for dp in sim:
+
+            #color arbiters
+            #order consumer based on arbiters
+            self.color_arb_blocks(dp, ctcr, arbiters)
+            self.color_con_blocks(dp, ctcr)
+
+        self.offset_row += 1
+
+    #make a loop that will color in each block depending on the line
+
+    def color_arb_blocks(self, dp, ctcr, arbiters):
+
+        if isinstance(dp, ArbiterDataPoint):
+
+            values = ctcr.get(dp.consumer_id)
```

Comment:
You can pack the values tuple immediately. Ex. If the map contains tuples (color1, color2, row) as values, this code works
```python
c1, c2, r = ctcr.get(dp.consumer_id)
```